File: gosdk/pre.go
package gosdk

import (
	"crypto/sha256"
	"crypto/sha512"

	"github.com/ChainSafe/go-schnorrkel"
	"github.com/gtank/ristretto255"
	"github.com/pkg/errors"
)

type Capsule struct {
	E *ristretto255.Element
	V *ristretto255.Element
	S *ristretto255.Scalar
}

func GenPreKey(pkA *schnorrkel.PublicKey) (*Capsule, []byte, error) {
	// generate E,V key-pairs
	skE, pkE, err := schnorrkel.GenerateKeypair()
	if err != nil {
		return nil, nil, errors.Wrap(err, "generate key error")
	}
	skV, pkV, err := schnorrkel.GenerateKeypair()
	if err != nil {
		return nil, nil, errors.Wrap(err, "generate key error")
	}
	// get H(E || V)
	ebpk := pkToElement(pkE).Encode([]byte{})
	vbpk := pkToElement(pkV).Encode([]byte{})
	h := hashAndConvToScalar(append(ebpk[:], vbpk[:]...))
	// get s = v + e * H(E || V)
	s := ristretto255.NewScalar().Add(
		skToScalar(skV),
		ristretto255.NewScalar().Multiply(skToScalar(skE), h),
	)
	// get (pk_A)^{e+v}
	point := pkToElement(pkA)
	point.ScalarMult(
		ristretto255.NewScalar().Add(
			skToScalar(skE),
			skToScalar(skV),
		),
		point,
	)
	// gen AES key
	txtBytes, err := point.MarshalText()
	if err != nil {
		return nil, nil, errors.Wrap(err, "generate key error")
	}
	key := sha256.Sum256(txtBytes)

	return &Capsule{
		E: pkToElement(pkE),
		V: pkToElement(pkV),
		S: s,
	}, key[:], nil
}

func DecryptKey(sk *schnorrkel.SecretKey, capsule *Capsule) ([]byte, error) {
	// get (pk_A)^{e+v}
	point := ristretto255.NewElement().ScalarMult(
		skToScalar(sk),
		ristretto255.NewElement().Add(capsule.E, capsule.V),
	)
	// gen AES key
	txtBytes, err := point.MarshalText()
	if err != nil {
		return nil, errors.Wrap(err, "decrypt key error")
	}
	key := sha256.Sum256(txtBytes)

	return key[:], nil
}

func GenReKey(skA *schnorrkel.SecretKey, pkB *schnorrkel.PublicKey) (*ristretto255.Scalar, *schnorrkel.PublicKey, error) {
	// generate x,X key-pair
	skX, pkX, err := schnorrkel.GenerateKeypair()
	if err != nil {
		return nil, nil, errors.Wrap(err, "generate re-encryption key error")
	}
	// get d=H(X||pk_B||pk_B^{x})
	xbpk, bbpk := pkX.Encode(), pkB.Encode()
	point := ristretto255.NewElement().ScalarMult(skToScalar(skX), pkToElement(pkB))
	txt, err := point.MarshalText()
	if err != nil {
		return nil, nil, errors.Wrap(err, "generate re-encryption key error")
	}
	d := hashAndConvToScalar(append(append(xbpk[:], bbpk[:]...), txt...))
	rk := ristretto255.NewScalar().Multiply(skToScalar(skA), d.Invert(d))
	return rk, pkX, nil
}

func ReEncryptKey(rk *ristretto255.Scalar, capsule *Capsule) (*Capsule, error) {
	// check sG==V+H(E||V)E
	sG := ristretto255.NewElement().ScalarBaseMult(capsule.S)
	ebpk := capsule.E.Encode([]byte{})
	vbpk := capsule.V.Encode([]byte{})
	h := hashAndConvToScalar(append(ebpk[:], vbpk[:]...))
	point := ristretto255.NewElement().Add(
		capsule.V,
		ristretto255.NewElement().ScalarMult(h, capsule.E),
	)
	if point.Equal(sG) != 1 {
		return nil, errors.Wrap(errors.New("invalid params"), "re-encrypt key error")
	}
	txt, err := capsule.S.MarshalText()
	if err != nil {
		return nil, errors.Wrap(err, "re-encrypt key error")
	}
	newS := ristretto255.NewScalar()
	if err = newS.UnmarshalText(txt); err != nil {
		return nil, errors.Wrap(err, "re-encrypt key error")
	}
	return &Capsule{
		E: ristretto255.NewElement().ScalarMult(rk, capsule.E),
		V: ristretto255.NewElement().ScalarMult(rk, capsule.V),
		S: newS,
	}, nil
}

func DecryptReKey(skB *schnorrkel.SecretKey, newCapsule *Capsule, pkX *schnorrkel.PublicKey) ([]byte, error) {
	// S= pkX^{skB}
	S := ristretto255.NewElement().ScalarMult(skToScalar(skB), pkToElement(pkX))
	txt, err := S.MarshalText()
	if err != nil {
		return nil, errors.Wrap(err, "decrypt re-encryption key error")
	}
	// recreate d=H(pkX || pkB || S)
	pkB, err := skB.Public()
	if err != nil {
		return nil, errors.Wrap(err, "decrypt re-encryption key error")
	}
	xbpk, bbpk := pkX.Encode(), pkB.Encode()
	d := hashAndConvToScalar(append(append(xbpk[:], bbpk[:]...), txt...))
	point := ristretto255.NewElement().ScalarMult(
		d,
		ristretto255.NewElement().Add(
			newCapsule.E,
			newCapsule.V,
		),
	)
	txtBytes, err := point.MarshalText()
	if err != nil {
		return nil, errors.Wrap(err, "decrypt re-encryption key error")
	}
	key := sha256.Sum256(txtBytes)
	return key[:], nil
}

func skToScalar(sk *schnorrkel.SecretKey) *ristretto255.Scalar {
	s := ristretto255.NewScalar()
	bs := sk.Encode()
	s.Decode(bs[:])
	return s
}

func pkToElement(pk *schnorrkel.PublicKey) *ristretto255.Element {
	e := ristretto255.NewElement()
	bpk := pk.Encode()
	e.Decode(bpk[:])
	return e
}

func hashAndConvToScalar(data []byte) *ristretto255.Scalar {
	hash := sha512.Sum512(data)
	return ristretto255.NewScalar().FromUniformBytes(hash[:])
}

File: rssdk/src/pre.rs
use curve25519_dalek::constants::RISTRETTO_BASEPOINT_POINT;
use curve25519_dalek::ristretto::{CompressedRistretto, RistrettoPoint};
use curve25519_dalek::scalar::Scalar;
use rand::rngs::OsRng;
use schnorrkel::{Keypair, PublicKey, SecretKey};
use sha2::{Digest, Sha256, Sha512};
use std::error::Error;

pub struct Capsule {
    pub e: RistrettoPoint,
    pub v: RistrettoPoint,
    pub s: Scalar,
}

/// Generate a pre-key capsule and AES key
pub fn gen_pre_key(pk_a: &PublicKey) -> Result<(Capsule, [u8; 32]), Box<dyn Error>> {
    let mut csprng = OsRng;

    let keypair_e = Keypair::generate_with(&mut csprng);
    let keypair_v = Keypair::generate_with(&mut csprng);

    let e_point = ristretto_point_from_pk(&keypair_e.public);
    let v_point = ristretto_point_from_pk(&keypair_v.public);

    let mut hasher = Sha512::new();
    hasher.update(e_point.compress().as_bytes());
    hasher.update(v_point.compress().as_bytes());
    let h_scalar = Scalar::from_hash(hasher);

    // point = pk_a^(e+v)
    let secret_bytes = keypair_v.secret.to_bytes();
    let scalar_v = Scalar::from_bytes_mod_order(
        secret_bytes
            .get(0..32)
            .ok_or("Invalid secret key length for v")?
            .try_into()?,
    );

    let secret_bytes = keypair_e.secret.to_bytes();
    let scalar_e = Scalar::from_bytes_mod_order(
        secret_bytes
            .get(0..32)
            .ok_or("Invalid secret key length for e")?
            .try_into()?,
    );

    let s = scalar_v + scalar_e * h_scalar;
    let scalar_sum = scalar_e + scalar_v;

    let pk_point = ristretto_point_from_pk(pk_a);
    let shared_point = scalar_sum * pk_point;

    let mut hasher = Sha256::new();
    hasher.update(shared_point.compress().as_bytes());
    let key = hasher.finalize();

    Ok((
        Capsule {
            e: e_point,
            v: v_point,
            s,
        },
        key.into(),
    ))
}

/// Decrypt a capsule using the secret key
pub fn decrypt_key(sk: &SecretKey, capsule: &Capsule) -> Result<[u8; 32], Box<dyn Error>> {
    let secret_bytes = sk.to_bytes();
    let sk_scalar = Scalar::from_bytes_mod_order(
        secret_bytes
            .get(0..32)
            .ok_or("Invalid secret key length: expected 32 bytes")?
            .try_into()?,
    );

    let sum = capsule.e + capsule.v;
    let point = sk_scalar * sum;

    let mut hasher = Sha256::new();
    hasher.update(point.compress().as_bytes());
    let key = hasher.finalize();

    Ok(key.into())
}

/// Generate a re-encryption key for skA -> pkB
pub fn gen_re_key(
    sk_a: &SecretKey,
    pk_b: &PublicKey,
) -> Result<(Scalar, PublicKey), Box<dyn Error>> {
    let mut csprng = OsRng;

    // Generate x,X key-pair
    let keypair_x = Keypair::generate_with(&mut csprng);
    let pk_x = keypair_x.public;
    let sk_x_scalar = Scalar::from_bytes_mod_order(keypair_x.secret.to_bytes()[0..32].try_into()?);

    // Compute d = H(X || pk_B || pk_B^x)
    let pk_b_point = ristretto_point_from_pk(pk_b);
    let point = sk_x_scalar * pk_b_point;

    let mut hasher = Sha512::new();
    hasher.update(pk_x.to_bytes());
    hasher.update(pk_b.to_bytes());
    hasher.update(point.compress().as_bytes());
    let d = Scalar::from_hash(hasher);

    // rk = skA * d^-1
    let sk_a_scalar = Scalar::from_bytes_mod_order(sk_a.to_bytes()[0..32].try_into()?);
    let rk = sk_a_scalar * d.invert();

    Ok((rk, pk_x))
}

/// Re-encrypt a capsule using a re-encryption key
pub fn re_encrypt_key(rk: &Scalar, capsule: &Capsule) -> Result<Capsule, Box<dyn Error>> {
    // sG = S * basepoint
    let s_g = capsule.s * RISTRETTO_BASEPOINT_POINT;

    // h = H(E || V)
    let mut hasher = Sha512::new();
    hasher.update(capsule.e.compress().as_bytes());
    hasher.update(capsule.v.compress().as_bytes());
    let h = Scalar::from_hash(hasher);

    // point = V + h * E
    let point = capsule.v + h * capsule.e;

    // Verify capsule
    if point != s_g {
        return Err("invalid params: re-encrypt key verification failed".into());
    }

    // Re-encrypt: E' = rk * E, V' = rk * V
    let new_e = rk * capsule.e;
    let new_v = rk * capsule.v;

    // S stays the same
    let new_s = capsule.s;

    Ok(Capsule {
        e: new_e,
        v: new_v,
        s: new_s,
    })
}

/// Decrypt a re-encrypted capsule using skB and ephemeral public key pkX
pub fn decrypt_re_key(
    sk_b: &SecretKey,
    new_capsule: &Capsule,
    pk_x: &PublicKey,
) -> Result<[u8; 32], Box<dyn Error>> {
    // S = pkX ^ skB
    let sk_b_scalar = Scalar::from_bytes_mod_order(sk_b.to_bytes()[0..32].try_into()?);
    let pk_x_point = ristretto_point_from_pk(pk_x);
    let s_point = sk_b_scalar * pk_x_point;

    // Get pkB from skB
    let pk_b = sk_b.to_public();

    // d = H(pkX || pkB || S)
    let mut hasher = Sha512::new();
    hasher.update(pk_x.to_bytes());
    hasher.update(pk_b.to_bytes());
    hasher.update(s_point.compress().as_bytes());
    let d = Scalar::from_hash(hasher);

    // point = d * (E + V)
    let sum = new_capsule.e + new_capsule.v;
    let point = d * sum;

    // AES key = SHA256(point)
    let mut sha_hasher = Sha256::new();
    sha_hasher.update(point.compress().as_bytes());
    let key = sha_hasher.finalize();

    Ok(key.into())
}

/// Helper: Convert PublicKey to RistrettoPoint
fn ristretto_point_from_pk(pk: &PublicKey) -> RistrettoPoint {
    let compressed = CompressedRistretto(pk.to_bytes());
    compressed.decompress().expect("Invalid public key")
}

#[cfg(test)]
mod tests {
    use super::*;
    use hex;
    use rand::rngs::OsRng;
    use schnorrkel::Keypair;

    #[test]
    fn test_proxy_re_encryption() -> Result<(), Box<dyn std::error::Error>> {
        let mut csprng = OsRng;

        // Generate keypairs for Alice (A), Bob (B), and Carol (C)
        let keypair_a = Keypair::generate_with(&mut csprng);
        let keypair_b = Keypair::generate_with(&mut csprng);
        let keypair_c = Keypair::generate_with(&mut csprng);

        let sk_a = &keypair_a.secret;
        let pk_a = &keypair_a.public;

        let sk_b = &keypair_b.secret;
        let pk_b = &keypair_b.public;

        let sk_c = &keypair_c.secret;

        // Generate pre-key capsule and AES key for Alice
        let (capsule, aes_key) = gen_pre_key(pk_a)?;
        println!("AES key: {}", hex::encode(aes_key));

        // Decrypt capsule with Alice's key
        let de_aes_key = decrypt_key(sk_a, &capsule)?;
        println!("decrypt AES key: {}", hex::encode(de_aes_key));

        // Decrypt capsule with Carol's key (fake)
        let fde_aes_key = decrypt_key(sk_c, &capsule)?;
        println!("decrypt AES key with fake sk: {}", hex::encode(fde_aes_key));

        // Generate re-encryption key for Alice -> Bob
        let (rk, x_pk) = gen_re_key(sk_a, pk_b)?;

        // Re-encrypt the capsule
        let new_capsule = re_encrypt_key(&rk, &capsule)?;

        // Decrypt re-encrypted capsule with Bob
        let dr_aes_key = decrypt_re_key(sk_b, &new_capsule, &x_pk)?;
        println!("decrypt re-encryption AES key: {}", hex::encode(dr_aes_key));

        // Decrypt re-encrypted capsule with Carol (fake)
        let fde_re_aes_key = decrypt_re_key(sk_c, &new_capsule, &x_pk)?;
        println!(
            "decrypt re-encryption AES key with fake skB: {}",
            hex::encode(fde_re_aes_key)
        );

        Ok(())
    }
}
